resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala

```diff
@@ -0,0 +1,234 @@
+ * watcher that monitors and logs the application status. Waits for the application to terminate if
+ * spark.kubernetes.submission.waitAppCompletion is true.
+ *
+ * @param submissionSteps steps that collectively configure the driver
+ * @param submissionSparkConf the submission client Spark configuration
+ * @param kubernetesClient the client to talk to the Kubernetes API server
+ * @param waitForAppCompletion a flag indicating whether the client should wait for the application
+ *                             to complete
+ * @param appName the application name
+ * @param loggingPodStatusWatcher a watcher that monitors and logs the application status
+ */
+private[spark] class Client(
+    submissionSteps: Seq[DriverConfigurationStep],
+    submissionSparkConf: SparkConf,
+    kubernetesClient: KubernetesClient,
+    waitForAppCompletion: Boolean,
+    appName: String,
+    loggingPodStatusWatcher: LoggingPodStatusWatcher) extends Logging {
+
+  private val driverJavaOptions = submissionSparkConf.get(
+    org.apache.spark.internal.config.DRIVER_JAVA_OPTIONS)
+
+   /**
+    * Run command that initializes a DriverSpec that will be updated after each
+    * DriverConfigurationStep in the sequence that is passed in. The final KubernetesDriverSpec
+    * will be used to build the Driver Container, Driver Pod, and Kubernetes Resources
+    */
+  def run(): Unit = {
+    var currentDriverSpec = KubernetesDriverSpec.initialSpec(submissionSparkConf)
+    // submissionSteps contain steps necessary to take, to resolve varying
+    // client arguments that are passed in, created by orchestrator
+    for (nextStep <- submissionSteps) {
+      currentDriverSpec = nextStep.configureDriver(currentDriverSpec)
+    }
+
+    val resolvedDriverJavaOpts = currentDriverSpec
+      .driverSparkConf
+      // Remove this as the options are instead extracted and set individually below using
+      // environment variables with prefix SPARK_JAVA_OPT_.
+      .remove(org.apache.spark.internal.config.DRIVER_JAVA_OPTIONS)
+      .getAll
+      .map {
+        case (confKey, confValue) => s"-D$confKey=$confValue"
+      } ++ driverJavaOptions.map(Utils.splitCommandString).getOrElse(Seq.empty)
+    val driverJavaOptsEnvs: Seq[EnvVar] = resolvedDriverJavaOpts.zipWithIndex.map {
+      case (option, index) =>
+        new EnvVarBuilder()
+          .withName(s"$ENV_JAVA_OPT_PREFIX$index")
+          .withValue(option)
+          .build()
+    }
+
+    val resolvedDriverContainer = new ContainerBuilder(currentDriverSpec.driverContainer)
+      .addAllToEnv(driverJavaOptsEnvs.asJava)
+      .build()
+    val resolvedDriverPod = new PodBuilder(currentDriverSpec.driverPod)
+      .editSpec()
+        .addToContainers(resolvedDriverContainer)
+        .endSpec()
+      .build()
+
+    Utils.tryWithResource(
+        kubernetesClient
+          .pods()
+          .withName(resolvedDriverPod.getMetadata.getName)
+          .watch(loggingPodStatusWatcher)) { _ =>
```

Comment:
In the interest of reducing the complexity of this particular change, I think it's fine to remove the logging pod status watcher for now and introduce it in a follow up commit. We can make a simpler version here. For example, we can have a simpler watcher that checks the pod ends up in the running state before some timeout. We can also explicitly not support waiting for app completion.

Basically I'm curious as to how simple we can make this to minimize the amount of new code we're introducing in the first push. The tradeoff of course is that we're expiicitly limiting functionality we introduce in 2.3.

Reply:
I personally don't think we can reduce much complexity as what are included in this PR and the first one form a MVP that support the most basic use cases: allowing users to submit Spark applications to a Kubernetes cluster with `local://` dependencies baked into images. Also given that this has already gone through several rounds of reviews by different reviewers, we are probably at a better position if we keep the current form. The next PR most probably will be for integration tests so we get more confident on the health of the MVP.

Reply:
What are your thoughts of breaking up the change in how I suggest in https://github.com/apache/spark/pull/19717#pullrequestreview-81034043? We can have some PRs that create unusable code but the combination of all of said PRs create the final product.

Reply:
Given the current state of this PR, I believe it's not far from being OK to merge. Breaking this up into multiple ones likely will not make the time to merge shorter, and possibly even make them harder to review because of lack of contexts in a single PR. IMO, I think we should keep this PR in its current form and dedicate the next one to integration tests. @felixcheung @foxish WDYT?